diff --git a/folly/Conv.h b/folly/Conv.h
index a586751f2..f298a5ff6 100644
--- a/folly/Conv.h
+++ b/folly/Conv.h
@@ -117,7 +117,7 @@
 #include <type_traits>
 #include <utility>

-#include <double-conversion/double-conversion.h> // V8 JavaScript implementation
+#include "double-conversion/double-conversion.h" // V8 JavaScript implementation

 #include <folly/CPortability.h>

diff --git a/folly/Format.cpp b/folly/Format.cpp
index 67c2efffa..2c97916d6 100644
--- a/folly/Format.cpp
+++ b/folly/Format.cpp
@@ -23,7 +23,7 @@
 #include <folly/Portability.h>
 #include <folly/container/Array.h>

-#include <double-conversion/double-conversion.h>
+#include "double-conversion/double-conversion.h"

 namespace folly {
 namespace detail {
